### ISSUE TITLE
Ping causing homebridge to crash

### ISSUE THREAD

**Issue by Nastras** [lprhodes/homebridge-broadlink-rm#325]
Hey Luke,
i have updated to version 3.3.3. I get this error every few hours with a reboot of the homebridge.

![72f04613-4adb-47b6-8460-3fa2f2274f97](https://user-images.githubusercontent.com/27686185/37786069-ff2cc6cc-2dfb-11e8-87db-96aebc7ffbb0.jpeg)

**Comment by lprhodes** [comment]
Can you send through the log rather than a screenshot please

**Comment by lprhodes** [comment]
Also, do you mean it's rebooting homebridge or homebridge reboots then this happens?

**Comment by Nastras** [comment]
I meant "restart" the homebridge broadlink instance. I have isolated each plugin in its own instance.
The Broadlink instance breaks off with the failure and restarts

I looked at the full log and found this from tonight:

![report](https://user-images.githubusercontent.com/27686185/37809235-9cfef03a-2e4e-11e8-9495-29d45236e7df.jpeg)


besides, I noticed this here, I did not change the device. It's all the same.

![img_0316](https://user-images.githubusercontent.com/27686185/37809275-cdc10262-2e4e-11e8-8d7e-478f525db604.jpg)

Does this help you or do you need the whole log file? If so, would I like to email you?

**Comment by lprhodes** [comment]
Screenshots of logs make things hard; they often don't include all the necessary information and it makes it hard for other people to discover similar issues.

You can paste the logs into Github or attach it as a file.

**Comment by Nastras** [comment]
There is a lot of information about my home. I upload the file syslog briefly and you tell me when you got it that I can delete it again ok?

**Comment by Nastras** [comment]
[syslog.zip](https://github.com/lprhodes/homebridge-broadlink-rm/files/1840145/syslog.zip)

**Comment by Nastras** [comment]
Did you get it?

**Comment by lprhodes** [comment]
Yes

**Comment by lprhodes** [comment]
Are you:

A) automatically restarting homebridge and the ping error shows
B) the ping error shows and crashes homebridge which then restarts

**Comment by Nastras** [comment]
B

**Comment by Nastras** [comment]
At 2 o'clock, a backup was performed automatically. That's why all the things went down there.

And i made manually a reboot between 4 and 5

**Comment by lprhodes** [comment]
OK - I believe it's something very specific to your setup.

The plugin pings the device you add regularly to check that it's still there.

In order to do so is creates a new `ping` child process and collects the information from the output.

The error is showing that there's no standard output pipe for the output to be fed into (stdout)

Rename this file to `ping-promise.js` and put replace the original with it in `node_modules/ping/lib`
[ping-promise.txt](https://github.com/lprhodes/homebridge-broadlink-rm/files/1840188/ping-promise.txt)

Then start homebridge. Then we should get some more information.

**Comment by Nastras** [comment]
I have made. Now wait until the error has recurred again?

**Comment by lprhodes** [comment]
Yeah

**Comment by Nastras** [comment]
☕️☕️☕️😄👍

**Comment by SeydX** [comment]
i'm getting the same error which kills homebridge every few hours

`Mär 23 06:13:16 raspberrypi homebridge[31901]: /usr/local/lib/node_modules/homebridge-broadlink-rm/node_modules/ping/lib/ping-promise.js:66
Mär 23 06:13:16 raspberrypi homebridge[31901]:     ping.stdout.on('data', function (data) {
Mär 23 06:13:16 raspberrypi homebridge[31901]:                 ^
Mär 23 06:13:16 raspberrypi homebridge[31901]: TypeError: Cannot read property 'on' of undefined`

**Comment by Nastras** [comment]
That's good, that means it's not just my setup only.

**Comment by tamasvicsotka** [comment]
Mar 23 11:13:30 pimaster homebridge[26489]: /usr/lib/node_modules/homebridge-broadlink-rm/node_modules/ping/lib/ping-promise.js:66
Mar 23 11:13:30 pimaster homebridge[26489]:     ping.stdout.on('data', function (data) {
Mar 23 11:13:30 pimaster homebridge[26489]:                ^
Mar 23 11:13:30 pimaster homebridge[26489]: TypeError: Cannot read property 'on' of undefined
Mar 23 11:13:30 pimaster homebridge[26489]:     at Object.probe (/usr/lib/node_modules/homebridge-broadlink-rm/node_modules/ping/lib/ping-promise.js:66:16)
Mar 23 11:13:30 pimaster homebridge[26489]:     at Object.probe (/usr/lib/node_modules/homebridge-broadlink-rm/node_modules/ping/lib/ping-sys.js:38:17)
Mar 23 11:13:30 pimaster homebridge[26489]:     at Timeout.setInterval [as _onTimeout] (/usr/lib/node_modules/homebridge-broadlink-rm/helpers/getDevice.js:10:14)
Mar 23 11:13:30 pimaster homebridge[26489]:     at ontimeout (timers.js:386:14)
Mar 23 11:13:30 pimaster homebridge[26489]:     at tryOnTimeout (timers.js:250:5)
Mar 23 11:13:30 pimaster homebridge[26489]:     at Timer.listOnTimeout (timers.js:214:5)
Mar 23 11:13:30 pimaster systemd[1]: homebridge.service: Main process exited, code=exited, status=1/FAILURE
Mar 23 11:13:30 pimaster systemd[1]: homebridge.service: Unit entered failed state.
Mar 23 11:13:30 pimaster systemd[1]: homebridge.service: Failed with result 'exit-code'.

**Comment by lprhodes** [comment]
I assume you've not seen this before the last couple of days?

If anyone else wants to do what I suggested with the ping-promise.js file then it may help speed up finding a solution.

**Comment by Nastras** [comment]
I just checked the SysLog. At 11:31 am the Homebridge crashed with broadlink-rm.

The only hint in the SysLog is this one:

Mar 23 11:26:10 HomeBridgeServer homebridge[603]: [2018-3-23 11:26:10] [Calendar] Pushing calendar sensor 'Kalendersensor Zu Hause' state 0 - value 0
Mar 23 11:26:10 HomeBridgeServer homebridge[603]: [2018-3-23 11:26:10] [Calendar] No events to schedule.
Mar 23 11:26:59 HomeBridgeServer homebridge[598]: [2018-3-23 11:26:59] [Router Abstellkammer] Updating guest WLAN
Mar 23 11:26:59 HomeBridgeServer homebridge[598]: [2018-3-23 11:26:59] [Router Abstellkammer] Getting guest WLAN state
Mar 23 11:27:59 HomeBridgeServer homebridge[598]: [2018-3-23 11:27:59] [Router Abstellkammer] Updating guest WLAN
Mar 23 11:27:59 HomeBridgeServer homebridge[598]: [2018-3-23 11:27:59] [Router Abstellkammer] Getting guest WLAN state
Mar 23 11:28:59 HomeBridgeServer homebridge[598]: [2018-3-23 11:28:59] [Router Abstellkammer] Updating guest WLAN
Mar 23 11:28:59 HomeBridgeServer homebridge[598]: [2018-3-23 11:28:59] [Router Abstellkammer] Getting guest WLAN state
Mar 23 11:29:59 HomeBridgeServer homebridge[598]: [2018-3-23 11:29:59] [Router Abstellkammer] Updating guest WLAN
Mar 23 11:29:59 HomeBridgeServer homebridge[598]: [2018-3-23 11:29:59] [Router Abstellkammer] Getting guest WLAN state
Mar 23 11:30:55 HomeBridgeServer homebridge[600]: [2018-3-23 11:30:55] [Thermostat Garage] Loading new data from API for: weatherstation
Mar 23 11:30:59 HomeBridgeServer homebridge[598]: [2018-3-23 11:30:59] [Router Abstellkammer] Updating guest WLAN
Mar 23 11:30:59 HomeBridgeServer homebridge[598]: [2018-3-23 11:30:59] [Router Abstellkammer] Getting guest WLAN state
Mar 23 11:31:10 HomeBridgeServer homebridge[603]: [2018-3-23 11:31:10] [Calendar] Updating calendar Kalendersensor Zu Hause
Mar 23 11:31:11 HomeBridgeServer homebridge[603]: [2018-3-23 11:31:11] [Calendar] Pushing calendar sensor 'Kalendersensor Zu Hause' state 0 - value 0
Mar 23 11:31:11 HomeBridgeServer homebridge[603]: [2018-3-23 11:31:11] [Calendar] No events to schedule.
Mar 23 11:31:11 HomeBridgeServer homebridge[617]: PING HAS NO stdout OR stderr
Mar 23 11:31:59 HomeBridgeServer homebridge[598]: [2018-3-23 11:31:59] [Router Abstellkammer] Updating guest WLAN
Mar 23 11:31:59 HomeBridgeServer homebridge[598]: [2018-3-23 11:31:59] [Router Abstellkammer] Getting guest WLAN state
Mar 23 11:32:59 HomeBridgeServer homebridge[598]: [2018-3-23 11:32:59] [Router Abstellkammer] Updating guest WLAN
Mar 23 11:32:59 HomeBridgeServer homebridge[598]: [2018-3-23 11:32:59] [Router Abstellkammer] Getting guest WLAN state
Mar 23 11:33:59 HomeBridgeServer homebridge[598]: [2018-3-23 11:33:59] [Router Abstellkammer] Updating guest WLAN
Mar 23 11:33:59 HomeBridgeServer homebridge[598]: [2018-3-23 11:33:59] [Router Abstellkammer] Getting guest WLAN state
Mar 23 11:34:59 HomeBridgeServer homebridge[598]: [2018-3-23 11:34:59] [Router Abstellkammer] Updating guest WLAN
Mar 23 11:34:59 HomeBridgeServer homebridge[598]: [2018-3-23 11:34:59] [Router Abstellkammer] Getting guest WLAN state
Mar 23 11:35:55 HomeBridgeServer homebridge[600]: [2018-3-23 11:35:55] [Thermostat Garage] Loading new data from API for: weatherstation
Mar 23 11:35:59 HomeBridgeServer homebridge[598]: [2018-3-23 11:35:59] [Router Abstellkammer] Updating guest WLAN
Mar 23 11:35:59 HomeBridgeServer homebridge[598]: [2018-3-23 11:35:59] [Router Abstellkammer] Getting guest WLAN state
Mar 23 11:36:11 HomeBridgeServer homebridge[603]: [2018-3-23 11:36:11] [Calendar] Updating calendar Kalendersensor Zu Hause

What amazes me is that Homebridge does not restart now, it's frozen like since I swapped files.

pi@HomeBridgeServer:~ $ sudo journalctl -f -au homebridge-broadlink-rm
-- Logs begin at Fri 2018-03-23 09:14:55 CET. --
Mär 23 09:17:43 HomeBridgeServer homebridge[617]: [2018-3-23 09:17:43] [Broadlink RM] Rollo 2 Wohnzimmer getTargetPosition: 100
Mär 23 09:18:09 HomeBridgeServer homebridge[617]: [2018-3-23 09:18:09] [Broadlink RM] Rollo 1 Wohnzimmer getCurrentPosition: 100
Mär 23 09:18:12 HomeBridgeServer homebridge[617]: [2018-3-23 09:18:12] [Broadlink RM] Rollo 3 Wohnzimmer getCurrentPosition: 100
Mär 23 09:18:30 HomeBridgeServer homebridge[617]: [2018-3-23 09:18:30] [Broadlink RM] Rollo 2 Schlafzimmer getPositionState: 2
Mär 23 10:49:42 HomeBridgeServer homebridge[617]: [2018-3-23 10:49:42] [Broadlink RM] Kerzen Wohnzimmer getSwitchState: false
Mär 23 10:49:42 HomeBridgeServer homebridge[617]: [2018-3-23 10:49:42] [Broadlink RM] Kerzen Wohnzimmer getBrightness: 100
Mär 23 10:49:42 HomeBridgeServer homebridge[617]: [2018-3-23 10:49:42] [Broadlink RM] Ambilight Wohnzimmer getSwitchState: undefined
Mär 23 10:49:59 HomeBridgeServer homebridge[617]: [2018-3-23 10:49:59] [Broadlink RM] Rollo 2 Schlafzimmer getPositionState: 2
Mär 23 10:50:06 HomeBridgeServer homebridge[617]: [2018-3-23 10:50:06] [Broadlink RM] Learn getSwitchState: undefined
Mär 23 11:31:11 HomeBridgeServer homebridge[617]: PING HAS NO stdout OR stderr

**Comment by lprhodes** [comment]
OK that's the output I needed thanks.

When did this start?

**Comment by Nastras** [comment]
The error started with V3.3.3 if I remember correctly.

**Comment by Nastras** [comment]
Can I replace the file or will it be overwritten by the next update?

**Comment by lprhodes** [comment]
You can update and it'll get replaced.

**Comment by lprhodes** [comment]
I've release 3.4.2 which should ignore the error when it occurs.

@Nastras I also noticed you're getting errors regarding unknown Homebridge devices. If you can narrow down the device and let me know which it is then I can handle it appropriately.

You have one with the device code `4620` and another with the code `4341`

**Comment by Nastras** [comment]
Because hint you mean I've seen so far only once.

I just searched for it in the log and can not find it again. Should he appear again I will open a new issue.

I'm testing the new version right now

**Comment by Nastras** [comment]
@lprhodes 

Now I get this error after a few hours:

Mar 23 20:58:02 HomeBridgeServer homebridge[603]: [2018-3-23 20:58:02] [Router Abstellkammer] Updating guest WLAN
Mar 23 20:58:02 HomeBridgeServer homebridge[603]: [2018-3-23 20:58:02] [Router Abstellkammer] Getting guest WLAN state
Mar 23 20:58:04 HomeBridgeServer homebridge[600]: os.js:126
Mar 23 20:58:04 HomeBridgeServer homebridge[600]:   const interfaceAddresses = getInterfaceAddresses();
Mar 23 20:58:04 HomeBridgeServer homebridge[600]:                              ^
Mar 23 20:58:04 HomeBridgeServer homebridge[600]: Error: EMFILE: too many open files, uv_interface_addresses
Mar 23 20:58:04 HomeBridgeServer homebridge[600]:     at Object.networkInterfaces (os.js:126:30)
Mar 23 20:58:04 HomeBridgeServer homebridge[600]:     at Broadlink.getIPAddresses (/usr/lib/node_modules/homebridge-broadlink-rm/node_modules/broadlinkjs-rm/index.js:70:27)
Mar 23 20:58:04 HomeBridgeServer homebridge[600]:     at Broadlink.discover (/usr/lib/node_modules/homebridge-broadlink-rm/node_modules/broadlinkjs-rm/index.js:57:30)
Mar 23 20:58:04 HomeBridgeServer homebridge[600]:     at Timeout.setInterval [as _onTimeout] (/usr/lib/node_modules/homebridge-broadlink-rm/helpers/getDevice.js:35:17)
Mar 23 20:58:04 HomeBridgeServer homebridge[600]:     at ontimeout (timers.js:466:11)
Mar 23 20:58:04 HomeBridgeServer homebridge[600]:     at tryOnTimeout (timers.js:304:5)
Mar 23 20:58:04 HomeBridgeServer homebridge[600]:     at Timer.listOnTimeout (timers.js:267:5)
Mar 23 20:58:05 HomeBridgeServer systemd[1]: homebridge-broadlink-rm.service: Main process exited, code=exited, status=1/FAILURE
Mar 23 20:58:05 HomeBridgeServer systemd[1]: homebridge-broadlink-rm.service: Unit entered failed state.
Mar 23 20:58:05 HomeBridgeServer systemd[1]: homebridge-broadlink-rm.service: Failed with result 'exit-code'.
Mar 23 20:58:15 HomeBridgeServer systemd[1]: homebridge-broadlink-rm.service: Service hold-off time over, scheduling restart.
Mar 23 20:58:15 HomeBridgeServer systemd[1]: Stopped homebridge-broadlink-rm.service.
Mar 23 20:58:15 HomeBridgeServer systemd[1]: Started homebridge-broadlink-rm.service.

**Comment by Nastras** [comment]
V 3.4.2

**Comment by lprhodes** [comment]
Thanks @Nastras 

I've implemented a fix based on a hunch. Can you try 3.4.3 please

**Comment by Nastras** [comment]
Just installed, now we have to wait again 😉

**Comment by lprhodes** [comment]
It’s been 5h so I’m hoping that means it’s not occurred?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by Nastras** [comment]
Until now not anymore, let's wait another 5 hours if the error has not surfaced, we close the issue. OK?

**Comment by lprhodes** [comment]
I’ll close it now and re-open if needed 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>
